raspberry/bzzz/read_sbus/read_sbus_from_GPIO_receiver.py

```diff
@@ -1,125 +1,139 @@
+        #     time.sleep(.2)
+
+        # Note that there will be nonsense data for the first 10ms or so of connection
+        # until the first packet comes in.
+        time.sleep(.1)
+        self.parser = bzzz.read_sbus.radioDataParser.RadioDataParser()
+
+        self.__parsed_data = None
+
+    def get_radio_data(self):
+        is_connected = self.reader.is_connected()
+        packet_age = self.reader.get_latest_packet_age()  # milliseconds
+
+        # returns list of length 16, so -1 from channel num to get index
+        channel_data = str(self.reader.translate_latest_packet())[1:-1]
+
+        return is_connected, packet_age, channel_data
+
+    def parse_radio_data(self, channel_data, over_write_throttle_ref_to=-1):
+        # check if data is in range [1000, 2000]
+        self.parser.m_channelData = list(map(lambda x: 0 if int(x) < 0 else (
+            2000 if int(x) > 2000 else int(x)), channel_data.strip().split(",")))
+
+        if over_write_throttle_ref_to != -1:
+            # TODO: check if channel 2 is throttle data
+            self.parser.m_channelData[2] = over_write_throttle_ref_to
+        # process and encapsulate the data
+        # the output data packet format will be as follows
+        # Y_radPs, P_rad, R_rad, T_PWM_MIN2MAX, % trimA, % trimB, % trimC, % trimE, encodedSwitchesData
+        # here the final data value encodedSwitchesData is an integer carrying information
+        # of the position of switches A, B, C, and D in the last 5-bits (the rightmost 5-bits),
+        # in which each bit corresponds to data as follows
+        # bit-4 (MSB): 1-bit info of Switch B: 1 if armed else 0
+        # bit-3: 1-bit info of Switch A: 1 if kill_on else 0
+        # bits 2 and 1: 2-bit info of switch C: 00 for position DOWN, 01 for position MID, 10 for position UP
+        # bit-0: 1-bit info of switch D: 1 if D_on else 0
+        self.__parsed_data, channel_data = self.parser.format_radio_data_for_sending()
+        return channel_data
+
+    def send_data_to_ESP(self, channel_data):
+        # Send with S in the beginning to indicate the start of the data, and also useful to check if data
+        # is received properly on the ESP's end
+        # Send data from Pi to ESP32, send a new line char so ESP32 knows when to stop reading
+        self.ser.write(f'S,{channel_data}\n'.encode())
+
+    def receive_data_from_ESP(self):
+        """Read data from ESP32 via UART.
+
+        :return: String if data is received, None otherwise.
+        """
+        while self.ser.inWaiting() > 0:
+            try:
+                line = self.ser.readline().decode('ascii').rstrip()
+                return line
+            except UnicodeDecodeError as e:
+                print(f"UnicodeDecodeError {e}, retrying....")
+        else:
+            return None
+
+    def get_radio_data_parse_and_send_to_ESP(self, return_channel_date=False, force_send_fake_data=False, fake_data="", over_write_throttle_ref_to=-1):
```

Comment:
`return_channel_date` should be `return_channel_data`?

Reply:
Done!